Add allDependencies method to LenientConfiguration

lhotari opened the pull request:
Adds convenience method for the complete set of dependencies including transitive dependencies

bmuschko commented on subprojects/core/src/main/java/org/gradle/api/artifacts/LenientConfiguration.java:
```diff
@@ -35,6 +35,13 @@
     /**
+     * Returns all successfully resolved dependencies.
+     *
+     * @return all resolved dependencies
```
Can you add a `@since` annotation to indicate the Gradle version that introduces the new method?

bmuschko commented on subprojects/core/src/main/java/org/gradle/api/artifacts/LenientConfiguration.java:
```diff
@@ -35,6 +35,13 @@
     public Set<ResolvedDependency> getFirstLevelModuleDependencies(Spec<? super Dependency> dependencySpec);
 
     /**
+     * Returns all successfully resolved dependencies.
```
Should mention that it also includes transitive dependencies.

bmuschko commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ErrorHandlingConfigurationResolver.java:
```diff
@@ -95,6 +95,15 @@ private ErrorHandlingLenientConfiguration(LenientConfiguration lenientConfigurat
+        public Set<ResolvedDependency> allDependencies() {
+            try {
+                return lenientConfiguration.allDependencies();
+            } catch (Throwable e){
```
Missing whitespace after parenthesis. Checkstyle might complain.

bmuschko commented on subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/DefaultLenientConfigurationTest.groovy:
```diff
@@ -0,0 +1,135 @@
+        Set<ResolvedArtifact> moduleArtifacts
+        Set<ResolvedArtifact> allModuleArtifacts
+
+        Set<ResolvedArtifact> getParentArtifacts(ResolvedDependency parent) {
```
Add `@Override` to align with usage for other methods.

bmuschko commented on subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/DefaultLenientConfigurationTest.groovy:
```diff
@@ -0,0 +1,135 @@
+        then:
+        1 * transientConfigurationResults.getRoot() >> root
+        result.size() == size
+        result.size() == expected.size()
```
Assertion probably not necessary as we are comparing expected result with the actual result.

bmuschko commented:
I am wondering if we still want to enhance `LenientConfiguration` even though you could probably just use [ResolutionResult.getAllDependencies()](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/result/ResolutionResult.html#getAllDependencies())? What's the concrete use case here?

bmuschko commented:
@lhotari Any chance you can look over my comments soon?

lhotari commented:
@bmuschko thanks for the review. I've rebased the PR branch and addressed the review comments. 

@jkschneider Please check that the API changes are what you were requesting in the hackathlon we had at the Gradle Summit. 

@bmuschko asked about the concrete use case: 
> I am wondering if we still want to enhance LenientConfiguration even though you could probably just use ResolutionResult.getAllDependencies()? What's the concrete use case here?

From what I understood, it's just about adding a convenience method for the complete set of dependencies including transitive dependencies. Since [ResolutionResult.getAllDependencies()](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/result/ResolutionResult.html#getAllDependencies()) returns a different type, I think that's not suitable for the use case.

bmuschko commented:
@lhotari The `ResolutionResult` API might be different but should provide the data/artifacts you need.

bmuschko commented:
Thanks for addressing my comments, @lhotari. I has been merged.

bmuschko commented:
Internal review: https://code-review.gradle.org/cru/REVIEW-6177

jkschneider commented:
@bmuschko For the sake of documenting this publicly for posterity, `ResolutionResult.getAllDependencies()` does not give us any way of accessing the binary artifacts. When we write plugins that need to read or manipulate the classpath in some way, this is a common need.
